docs/design/storagenode-downtime-tracking.md

```diff
@@ -0,0 +1,207 @@
+# Storage Node Downtime Tracking
+
+## Abstract
+
+The goal is tracking Storage Nodes downtime per Satellite.
+
+## Background
+
+The [Disqualification design document](disqualification.md) describes when the Storage Nodes get disqualified based on the reputation scores described in the [Node Selection design document](node-selection.md).
+
+The current disqualification showed to us that several Storage Nodes got disqualified because of their uptime stats without giving clear and fair evidence of it to us nor their operators. We had to take action and invalidate the uptime's part of the formula used for calculating their reputation.
+
+We need to enable back uptime disqualification, however before, we need to track in a fair way the amount of time Storage Nodes are offline.
+
+This document presents a solution for tracking  Storage Node offline time, but it's out of the scope of this document how Storage Node reputation is affected by it.
+
+The presented solution considers the requirements imposed by the [Kademlia removal design document](kademia-removal.md), where the Storage Nodes must contact the Satellite when they start and after every hour.
+
+## Design
+
+For tracking Storage Nodes offline time we need:
+
+- A new SQL database table to store offline time.
+- Detecting which ones are offline.
+- Roughly estimate how long they are offline.
+
+An _uptime check_ referenced in this section is the connection initiated by the Satellite to any Storage Node in the same way that's described in [Network refreshing section of the Kademlia removal design document](kademia-removal.md#network-refreshing).
+
+__NOTE__ the SQL code in this section is illustrative for explaining the algorithm concisely.
+
+### Database
+
+The following new SQL database table will store Storage Nodes offline time.
+
+```sql
+CREATE TABLE nodes_offline_time(
+    node_id BYTEA NOT NULL,
+    tracked_at timestamp with time zone NOT NULL,
+    seconds integer,                                                    -- Measured number of seconds being offline
+    CONSTRAINT nodes_offline_time__pk PRIMARY KEY(node_id, tracked_at)
+)
+```
+
+The current Satellite database has the table `nodes`. For the offline time calculation we use the following columns:
+
+- `last_contact_success`
+- `last_contact_failure`
+- `total_uptime_count`
+- `uptime_success_count`
+
+### Detecting offline nodes
+
+An independent process runs, in a configurable interval, of time the following query:
+
+```sql
+SELECT
+FROM nodes
+WHERE
+    last_contact_success < (now()  - 1h) AND
+    last_contact_success > last_contact_failure AND
+    disqualified IS NULL
+ORDER BY
+    last_contact_success ASC
+LIMIT 1
```

Comment:
The limit 1 in combination with the order is a build in killswitch. Lets say this query runs in an interval of 30s. Per day the satellite can track about 3000 storage node. As cheater I don't want that the satellite notice my downtime. All I have to do is creating 3000 fake nodes. The job of the fake nodes is to get selected first. They will ignore all other calls except that one important ping message. If they answer that ping they will get a last_contact_success renew and 0 downtime. At the end the satellite will cycle through the same 3000 fake nodes all the time. With that cycle the satellite my cheater node can stay offline for 1 day without any penalty. By increasing the number of fake nodes I can increase the timespan my cheater node can stay offline.

The issue here is scaling. I used an example with fake nodes but in production this might happen without any malicious intention. The big question is why do we want to limit the query to 1? Why don't we ping them all?

Reply:
We can ping them all as we can.
There is a limit of 1 for avoiding to hold the result set for a long time, but if that's not an issue then we can run without the limit.

I commented that in https://github.com/storj/storj/pull/2884#discussion_r318168100

Reply:
The limit will be removed because there is no DB performance issue.

Reply:
It has been removed.